### DISCUSSION TITLE
using slang-netlist as library?

### DISCUSSION THREAD

**Discussion by d0t2** [MikePopoloski/slang#1130] [Q&A]
Hi, I am trying to use the code in slang-netlist in my project. Specifically, I am using it to find all instances of "read" and "write" accesses to a variable. Which I believe can be done by retrieving the declaration netlist node and gathering its incoming and outgoing edges.

This is what my CMakeLists.txt file looks like right now:
```cmake
cmake_minimum_required(VERSION 3.20)
project(svrra)

include(FetchContent)

FetchContent_Declare(
  slang
  GIT_REPOSITORY https://github.com/MikePopoloski/slang.git
  GIT_SHALLOW ON)
FetchContent_MakeAvailable(slang)

FetchContent_Declare(
  yaml-cpp
  GIT_REPOSITORY https://github.com/jbeder/yaml-cpp.git
  GIT_SHALLOW ON)
FetchContent_MakeAvailable(yaml-cpp)

set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

add_executable(svrra main.cpp)
target_link_libraries(svrra PUBLIC slang::slang)
target_link_libraries(svrra PUBLIC yaml-cpp::yaml-cpp)

target_include_directories(svrra PRIVATE ${slang_SOURCE_DIR}/tools/netlist/include)
target_sources(svrra PRIVATE
    ${slang_SOURCE_DIR}/tools/netlist/source/Netlist.cpp
    ${slang_SOURCE_DIR}/tools/netlist/source/CombLoops.cpp)
```

With this, I am able to include files from tools/netlist/include, and it does build successfully, however when I try to create a netlist from an SV file with a concurrent assertion I get this error:

```
terminate called after throwing an instance of 'slang::assert::AssertionException'
  what():  Assertion 'T::isKind(kind)' failed
  in file [redacted]/svrra/build/_deps/slang-src/source/../include/slang/ast/Statements.h, line 238
  function: const T& slang::ast::Statement::as() const [with T = slang::ast::TimedStatement]

Abort
```

Trying to use a debugger to inspect this failed miserably. The error is odd to me because from the commit history, it looks like ignoring concurrent assertions was something that was specifically fixed in a recent commit. And indeed the same SV source file does work with the slang-netlist executable.

What's going wrong? How can I resolve this?

Unrelated note:
As I dig deeper into this project, I am more and more impressed by its sheer size. As someone who is constantly going through the IEEE SystemVerilog spec book, the fact that a single person can implement its ideas to any level of completion is mind-blowing to me, let alone a basically fully accurate implementation. Big props to you @MikePopoloski, the fact that you even seem to have a decorated career outside of this project is insane.

**Comment by MikePopoloski** [comment]
Thanks for the kind words.

The netlist tool isn't necessarily intended to be used as a library, but I suppose the way you're pulling it in will work. I'm not sure about the particulars of your question; there's no enough information to diagnose it. You'll need to look at the backtrace from when the assertion fires and see what the code is looking at. Your code might be configuring the tool differently from what slang-netlist will do by default. You can look at what the code is doing and try to replicate or figure out what you're doing differently when setting it up.

**Reply by d0t2** [reply]
I've attached my code below (sorry if there's any egregious style choices, this is a prototype and I'm learning C++ as I do this). I'm basically copying the way Netlist is initialized in netlist.cpp, I've also tried replacing the contents of this file with netlist.cpp and running that gave me the same error. Which is why I thought the error is caused by my CMake configs.

```cpp
#include <sstream>
#include <fstream>
#include "yaml-cpp/yaml.h"
#include "slang/ast/Compilation.h"
#include "slang/driver/Driver.h"
#include "slang/syntax/SyntaxVisitor.h"
#include "slang/syntax/SyntaxPrinter.h"
#include "Netlist.h"
#include "visitors/NetlistVisitor.h"

using namespace slang;

class RRARewriter : public syntax::SyntaxRewriter<RRARewriter> {
protected:
    std::unique_ptr<ast::Compilation> compilation;
    // [redacted]
    netlist::Netlist netlist;
    netlist::NetlistVisitor netlist_visitor;

public:
    explicit RRARewriter(
        std::unique_ptr<ast::Compilation> compil,
        // [redacted]
    ):
        syntax::SyntaxRewriter<RRARewriter>(),
        compilation(std::move(compil)),
        // [redacted]
        netlist_visitor(*compilation, netlist)
    {
        compilation->getRoot().visit(netlist_visitor);
    }

    // [redacted]
};

int main(int argc, char* argv[]) {
    YAML::Node config = YAML::LoadFile(argv[1]);

    std::ostringstream cmd_oss;
    // [redacted]
    cmd_oss << "svrra ";
    for (const YAML::Node& src_file : config["source_files"]) {
        cmd_oss << src_file.as<std::string>() << " ";
    }
    // [redacted]

    driver::Driver driver;
    driver.addStandardArgs();
    if (!driver.parseCommandLine(cmd_oss.str())) {
        return 1;
    }
    if (!driver.processOptions()) {
        return 2;
    }
    if (!driver.parseAllSources()) {
        return 3;
    }
    std::unique_ptr<ast::Compilation> compilation = driver.createCompilation();

    RRARewriter rewriter(std::move(compilation), std::move(array_paths), std::move(array_shapes));
    // [redacted]
    return 0;
}
```

This is the stack trace I got when I tried using VSCode's C++ debugger. Where ever I place the break point there's immediately an exception and I can't step through the program.

![image](https://github.com/user-attachments/assets/a5077f35-828e-4027-b754-e270d66f0e3d)

**Reply by d0t2** [reply]
Is there a way to include slang as a library with CMake such that its under the exact same conditions as when the slang-netlist executable is built?

**Reply by trile127** [reply]
Hmm i have this problem now as well

**Reply by MikePopoloski** [reply]
@jameshanlon what do you think about moving slang-netlist out of this repo and into its own downstream? There's a bunch of issues piling up and questions like this that make me think it ought to live on its to get proper support instead of being lost in the noise of this larger repo.

**Reply by jameshanlon** [reply]
Hi @MikePopoloski. Yes, I am open to this idea and have had the same thought recently. I have been refactoring the tool to make use of the new data flow analysis you have added recently, which would be more straightforward for me to complete in a separate repoistory. The only real disadvantage from my point of view is that having the netlist tool bundled with slang means that it is easy for people to find and ideally contribute to. I don't think this outweighs the benefits though. I'll raise a PR to remove it.

**Reply by MikePopoloski** [reply]
Thanks! FWIW I'm happy to put a link to it in the README. I should probably add a section for known users / downstream projects, so let me know when you have a repo to link to.

**Reply by jameshanlon** [reply]
@MikePopoloski Here's a link to the new home for slang-netlist:

https://github.com/jameshanlon/slang-netlist

That is, if you're still up for adding a link from this repository's documentation.

**Reply by MikePopoloski** [reply]
Ok yeah, I'll find a spot to highlight that in the docs.
